File: scripts/scrape.py
"""Scrape les expositions des sites de musées (sources hors « Que Faire à Paris »).

Chaque musée ayant un `expos_url` et un scraper enregistré est traité. Les
expositions trouvées sont insérées en `status="draft"` (file de validation admin),
rattachées au musée, avec source="scraping".

Usage :
  python scripts/scrape.py                 # dry-run : affiche, n'écrit rien
  python scripts/scrape.py --commit        # écrit en base (brouillons)
  python scripts/scrape.py --museum musee-du-louvre --commit
"""
import sys
import os
import argparse

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from slugify import slugify

from paname import create_app
from paname.extensions import db
from paname.models import Museum, Exposition
from paname.utils import find_duplicate_expo
from scripts.scrapers import SCRAPERS


def _unique_slug(base, ext_id):
    base = slugify(base)[:200] or "expo"
    slug, n = base, 2
    while True:
        clash = Exposition.query.filter(
            Exposition.slug == slug, Exposition.external_id != ext_id
        ).first()
        if not clash:
            return slug
        slug, n = f"{base}-{n}", n + 1


def upsert(museum, item):
    """Crée/maj une Exposition à partir d'un dict normalisé. Renvoie 'created'|'updated'."""
    ext_id = f"scrap:{museum.slug}:{item['ext_key']}"
    expo = Exposition.query.filter_by(external_id=ext_id).first()
    if not expo:   # déjà présente via une autre source (ex. Que Faire à Paris) ?
        expo = find_duplicate_expo(item["title"], museum.id, museum.name,
                                   item.get("date_start"))
    state = "updated"
    if not expo:
        expo = Exposition(
            external_id=ext_id,
            slug=_unique_slug(item["title"], ext_id),
            source="scraping",
            status="draft",          # passe par la validation admin
        )
        db.session.add(expo)
        state = "created"

    expo.title = item["title"]
    expo.description = item.get("description")
    expo.url = item.get("url")
    expo.image_url = item.get("image_url")
    expo.date_start = item.get("date_start")
    expo.date_end = item.get("date_end")
    expo.museum_id = museum.id
    expo.venue_name = museum.name
    expo.address = museum.address
    # Tarif de l'expo déduit des conditions de gratuité du musée.
    fa = museum.free_access_list
    if "gratuit_tous" in fa or "permanent" in fa:
        expo.price_category = "gratuit_tous"
    elif "gratuit_26" in fa:
        expo.price_category = "gratuit_26"
    else:
        expo.price_category = "gratuit_tous"
    return state


def run(only=None, commit=False):
    app = create_app()
    with app.app_context():
        targets = []
        for slug, fn in SCRAPERS.items():
            if only and slug != only:
                continue
            museum = Museum.query.filter_by(slug=slug).first()
            if not museum:
                print(f"⚠ Musée introuvable en base : {slug}")
                continue
            if not museum.expos_url:
                print(f"⚠ Pas d'expos_url renseignée pour {museum.name} — ignoré.")
                continue
            targets.append((museum, fn))

        if not targets:
            print("Aucun musée à scraper.")
            return

        for museum, fn in targets:
            print(f"\n→ {museum.name}  ({museum.expos_url})")
            try:
                items = fn(museum)
            except Exception as e:
                print(f"  ✗ Erreur scraping : {e}")
                continue
            print(f"  {len(items)} exposition(s) trouvée(s).")

            created = updated = 0
            for it in items:
                period = " → ".join(filter(None, [
                    it["date_start"].strftime("%d/%m/%Y") if it.get("date_start") else None,
                    it["date_end"].strftime("%d/%m/%Y") if it.get("date_end") else None,
                ])) or "dates ?"
                tags = ", ".join(it.get("tags", []))
                print(f"   • {it['title']}  [{period}]  {('— ' + tags) if tags else ''}")
                if commit:
                    state = upsert(museum, it)
                    created += state == "created"
                    updated += state == "updated"

            if commit:
                db.session.commit()
                print(f"  ✓ {created} créées (draft), {updated} mises à jour.")

        if not commit:
            print("\n(dry-run — rien n'a été écrit. Relance avec --commit pour enregistrer.)")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--museum", help="slug d'un seul musée à scraper")
    parser.add_argument("--commit", action="store_true", help="écrire en base")
    args = parser.parse_args()
    run(only=args.museum, commit=args.commit)

File: scripts/scrape.js
"use strict";

// Scrape les expositions des sites de musées (hors « Que Faire à Paris »).
// Insère en status="draft" (file de validation admin). Port de scripts/scrape.py.
//
// Usage :
//   npm run scrape                          # dry-run : affiche, n'écrit rien
//   npm run scrape -- --commit              # écrit en base (brouillons)
//   npm run scrape -- --museum musee-du-louvre --commit

const { Op } = require("sequelize");
const { sequelize, Museum, Exposition } = require("../src/models");
const { findDuplicateExpo } = require("../src/lib/dedup");
const { cleanValue, coerceUrl } = require("../src/lib/values");
const { toSlug } = require("../src/lib/slug");
const { SCRAPERS } = require("../src/scrapers");

function fmtDate(iso) {
  if (!iso) return null;
  const [y, m, d] = iso.split("-");
  return `${d}/${m}/${y}`;
}

async function uniqueSlug(title, extId) {
  const b = toSlug(title).slice(0, 200) || "expo";
  let slug = b;
  let n = 2;
  for (;;) {
    const clash = await Exposition.findOne({ where: { slug, external_id: { [Op.ne]: extId } } });
    if (!clash) return slug;
    slug = `${b}-${n}`;
    n += 1;
  }
}

async function upsert(museum, item) {
  const extId = `scrap:${museum.slug}:${item.ext_key}`;
  let expo = await Exposition.findOne({ where: { external_id: extId } });
  if (!expo) {
    expo = await findDuplicateExpo(Exposition, {
      title: item.title,
      museumId: museum.id,
      venueName: museum.name,
      dateStart: item.date_start,
    });
  }
  let state = "updated";
  if (!expo) {
    expo = Exposition.build({
      external_id: extId,
      slug: await uniqueSlug(item.title, extId),
      source: "scraping",
      status: "draft",
    });
    state = "created";
  }

  expo.title = item.title;
  expo.description = cleanValue(item.description);
  expo.url = coerceUrl(item.url);
  expo.image_url = coerceUrl(item.image_url);
  expo.date_start = item.date_start || null;
  expo.date_end = item.date_end || null;
  expo.museum_id = museum.id;
  expo.venue_name = cleanValue(museum.name);
  expo.address = cleanValue(museum.address);
  const fa = museum.free_access_list;
  if (fa.includes("gratuit_tous") || fa.includes("permanent")) expo.price_category = "gratuit_tous";
  else if (fa.includes("gratuit_26")) expo.price_category = "gratuit_26";
  else expo.price_category = "gratuit_tous";
  await expo.save();
  return state;
}

function parseArgs() {
  const args = process.argv.slice(2);
  const only = args.includes("--museum") ? args[args.indexOf("--museum") + 1] : null;
  const commit = args.includes("--commit");
  return { only, commit };
}

async function run() {
  await sequelize.authenticate();
  await sequelize.sync();
  const { only, commit } = parseArgs();

  const targets = [];
  for (const [slug, fn] of Object.entries(SCRAPERS)) {
    if (only && slug !== only) continue;
    const museum = await Museum.findOne({ where: { slug } });
    if (!museum) {
      console.log(`⚠ Musée introuvable en base : ${slug}`);
      continue;
    }
    if (!museum.expos_url) {
      console.log(`⚠ Pas d'expos_url pour ${museum.name} - ignoré.`);
      continue;
    }
    targets.push([museum, fn]);
  }

  if (!targets.length) {
    console.log("Aucun musée à scraper.");
    await sequelize.close();
    return;
  }

  for (const [museum, fn] of targets) {
    console.log(`\n→ ${museum.name}  (${museum.expos_url})`);
    let items;
    try {
      items = await fn(museum);
    } catch (e) {
      console.log(`  ✗ Erreur scraping : ${e.message}`);
      continue;
    }
    console.log(`  ${items.length} exposition(s) trouvée(s).`);

    let created = 0;
    let updated = 0;
    for (const it of items) {
      const period =
        [fmtDate(it.date_start), fmtDate(it.date_end)].filter(Boolean).join(" → ") || "dates ?";
      const tags = (it.tags || []).join(", ");
      console.log(`   • ${it.title}  [${period}]  ${tags ? "- " + tags : ""}`);
      if (commit) {
        const state = await upsert(museum, it);
        if (state === "created") created += 1;
        else updated += 1;
      }
    }
    if (commit) console.log(`  ✓ ${created} créées (draft), ${updated} mises à jour.`);
  }

  if (!commit) console.log("\n(dry-run - rien n'a été écrit. Relance avec --commit pour enregistrer.)");
  await sequelize.close();
}

run().catch((e) => {
  console.error("✗ Scrape échoué :", e);
  process.exit(1);
});
